### ISSUE TITLE
Volatile reads and writes on aarch64 sometimes generate instructions not suitable for MMIO in protected VMs

### ISSUE THREAD

**Issue by qwandor** [rust-lang/rust#131894]
`core::ptr::write_volatile` and `core::ptr::read_volatile` are documented as being intended to act on I/O memory, i.e. for MMIO. These are indeed widely used by many crates providing drivers for MMIO devices across the ecosystem.

When running in a virtual machine, MMIO must be emulated by the hypervisor. This is done (on aarch64 at least) by having the MMIO region unmapped in the stage 2 page table, which results in a data abort to the hypervisor when the VM attempts to read or write the MMIO region. The hypervisor then decodes the exception syndrome register (`esr_el2`) and uses the fault address register (`far_el2`) to determine which MMIO address is being accessed and perform the appropriate operation in response.

Unfortunately, rustc sometimes compiles `core::ptr::write_volatile` on aarch64 to something like `str w9, [x0], #4`. We've seen this happen particularly since Rust 1.78, but it may be possible with earlier Rust versions too. The problem with this is that this post-addressing mode is performing register writeback (in this case, incrementing `x0` by 4), and so [doesn't set the exception syndrome register](https://developer.arm.com/documentation/ddi0595/2020-12/AArch64-Registers/ESR-EL1--Exception-Syndrome-Register--EL1-?lang=en#fieldset_0-24_0_15-24_24). This prevents the hypervisor from emulating the MMIO access, as it has no way of decoding the instruction syndrome or finding the faulting address.

In an unprotected VM (e.g. regular KVM), it is possible for the VMM to work around this by reading the guest VM's memory to find the relevant instruction, decoding the instruction manually, and finding the MMIO address that way. This has a performance overhead and adds extra complexity. In the case of a protected VM where the host doesn't have access to the guest VM's memory (e.g. protected KVM), this is not possible as the VMM is not able to read the guest VM's memory and so cannot do instruction decoding. There is thus no way to emulate these attempted MMIO accesses in a protected VM on aarch64.

The net result of this is that instructions which perform register writeback (e.g. post-increment addressing modes) are not suitable for MMIO in aarch64 VMs. This is arguably a flaw in the aarch64 architecture, but as that's not feasible to fix at this point it must be fixed in the compiler instead. rustc should therefore avoid generating such instructions for `volatile_read` and `volatile_write` calls.

The only alternative I can see to fixing this in rustc is for every crate which performs MMIO to use inline assembly rather than `volatile_read` / `volatile_write`, but that is not a very feasible or scalable solution.

**Comment by Noratrieb** [comment]
Does clang make this work correctly with C `volatile`? If yes, how does rustc's assembly and LLVM IR differ from clang?

**Comment by workingjubilee** [comment]
> The only alternative I can see to fixing this in rustc is for every crate which performs MMIO to use inline assembly rather than `volatile_read` / `volatile_write`, but that is not a very feasible or scalable solution.

I wouldn't be so quick to assume this is not feasible or preferable, because if you want a very specific instruction, this might in fact be the preferred solution.

The reason why is this ain't gonna lower to a single `str` in *any* addressing mode. It **can't**:
```rust
ptr.cast::<[u8; 1024]>().write_volatile(bytes);
```

That is why I am very interested in seeing what Rust source code actually causes this to be a problem.

**Comment by maurer** [comment]
From @Noratrieb 
>Does clang make this work correctly with C volatile? If yes, how does rustc's assembly and LLVM IR differ from clang?

We have observed this eissue with `clang` as well in another context - this started happening b/c LLVM got smarter, and `clang` and `rustc` both had failing VMs as a result. Our solution in both cases so far has been to put an arch-specialized method into the code which does explicit inline asm to address this (arch-specific header for C, a `#[cfg]`'d trait implementation for Rust).

From @workingjubilee 
> The reason why is this ain't gonna lower to a single `str` in _any_ addressing mode:

The issue is not that it should lower to a single `str`, the problem is showing up in using `str` variants that have updates to additional registers - @qwandor 's example shows it doing a writeback to `x0` after the store to update the pointer. While it's true that this is still following normal volatile semantics (the write is not being elided), the usual `aarch64` VM MMIO implementation can't handle these (see the [syndrome register docs](https://developer.arm.com/documentation/ddi0595/2020-12/AArch64-Registers/ESR-EL1--Exception-Syndrome-Register--EL1-?lang=en#fieldset_0-24_0_15-24_24) linked earlier). This means that if someone tries to write an MMIO driver using `write_volatile`, it may mysteriously fail on ARM if it tries to generate a more efficient loop, because the hypervisor won't be able to read the fault address out of the syndrome register.

I see a few possible solutions to this:

1. Make `write_volatile` only generate MMIO compatible operations. On `aarch64`, this means restricting writeback. There might be other similar restrictions on other architectures.
2. Add a `write_mmio` method that is like `write_volatile`, but with the architecture-specific restrictions above. Given that it's for MMIO, it might even make sense to restrict it to word sizes for the relevant architecture, but that's not relevant to this specific issue.
3. As option 2, but do it in a support crate rather than in stdlib and implement via inline asm on archs that need a special restriction.
4. What we're doing now - each crate or mini-ecosystem that cares about this will encounter this and resolve it themselves when relevant.

**Comment by workingjubilee** [comment]
I mean, the Rust compiler could be the one emitting the inline asm when it monomorphizes a word-sized-or-less `write_volatile` instruction, for all it matters here.

**Comment by Noratrieb** [comment]
Given that clang has the same issue and rustc just compiles it to `load volatile`, this should be addressed on the LLVM side if possible.

**Comment by workingjubilee** [comment]
@qwandor @maurer Can this be reported as an LLVM bug, then, so we can find out if they want to fix it?

**Comment by the8472** [comment]
If LLVM doesn't do it, is this pattern even blessed by anyone? Does GCC guarantee that volatile writes are virtualizable?

**Comment by workingjubilee** [comment]
Minimal repro: https://godbolt.org/z/crGo6xzev

**Comment by jyn514** [comment]
for anyone here from the future: this is waiting on someone to report an LLVM bug upstream. Godbolt shows the following LLVM IR for Jubilee's repro:
```
define dso_local void @foo_rust(ptr noalias noundef nonnull align 4 %xs.0, i64 noundef %xs.1) unnamed_addr {
start:
  %end_or_len = getelementptr inbounds i32, ptr %xs.0, i64 %xs.1
  %0 = icmp eq i64 %xs.1, 0
  br i1 %0, label %bb5, label %bb4

bb5:
  ret void

bb4:
  %iter.sroa.0.03 = phi ptr [ %_30.i, %bb4 ], [ %xs.0, %start ]
  %_30.i = getelementptr inbounds i32, ptr %iter.sroa.0.03, i64 1
  store volatile i32 42, ptr %iter.sroa.0.03, align 4
  %1 = icmp eq ptr %_30.i, %end_or_len
  br i1 %1, label %bb5, label %bb4
}

declare void @llvm.dbg.value(metadata, metadata, metadata) #1
```
i think the relevant line is `store volatile i32 42, ptr %iter.sroa.0.03, align 4`.

**Comment by bjorn3** [comment]
> Volatile accesses are of dubious suitability for a large amount memory-mapped IO regardless

If that is RAM of another device you are accessing, you shouldn't need volatile. Plain atomic memory accesses (or even non-atomic ones if you can guarantee nobody else will touch it during the access) + suitable memory barriers and cache flushing operations (if non-coherent) would be an acceptable way I think. In this case you don't care that the compiler may insert spurious reads or deduplicate/merge writes. The accesses themselves don't have side effects. For actual MMIO registers on the other hand it does matter that the exact operation is preserved, rather than just the effect on the target memory. Which is exactly what volatile instructions do.

**Comment by workingjubilee** [comment]
And since the most likely place for that code would be the library,

@rustbot label +I-libs-nominated

**Comment by Darksonn** [comment]
I'm sure nobody made an active decision to decline your request to file an issue. I myself assumed one had already been filed when I discussed this with @nikic at RustWeek earlier this year.

**Comment by zjp-CN** [comment]
The problem also happens on x86_64: https://github.com/asterinas/asterinas/pull/2785#issuecomment-3798609076

---

> volatile is our recommended way to do MMIO and it ideally ought to work

But there is a proposal to have stricter volatile semantics 
* https://github.com/rust-lang/unsafe-code-guidelines/issues/321

**Comment by tatetian** [comment]
> [@zjp-CN](https://github.com/zjp-CN) To clarify, is that issue in asterinas also about the combination of MMIO plus a hypervisor/guest relationship, or does the generation of machine instructions incompatible with MMIO problem arise without such?

@workingjubilee Your understanding is correct. The problem occurs in a VM setting; or more precisely, in Confidential VMs such as Intel TDX and AMD SEV.

Basically, the [Asterinas](https://github.com/asterinas/asterinas) kernel used to do MMIO with `ptr::read_volatile` and `ptr::write_volatile` API. Our assumption was that such APIs would issue a single, naive memory load/store instruction (`mov` on x86). But in reality, we found that LLVM may do optimizations that merge a memory load/store with a subsequent operation, thereby issuing a more complex instruction (e.g., a `cmp` instruction on x86 if a memory load is followed by a comparison operation).

Such optimizations are problematic in the Confidential VM setting because every MMIO read/write triggers a hardware exception. To handle the exception, the kernel needs to _decode_ the faulting kernel instruction. But the instruction decoding ability of a kernel is limited: it can only decode a small set of anticipated instructions (e.g., `mov`). It is unacceptable for the compiler to do aggressive optimizations and generate arbitrary instructions for MMIO. Our workaround is to replace `ptr::read_volatile` and `ptr::write_volatile` with manually-written assembly code.

More info can be found in this [issue](https://github.com/asterinas/asterinas/issues/2948).

**Comment by workingjubilee** [comment]
@tatetian Thank you for that clarification! Can you open a new LLVM issue regarding this case? While superficially similar, it concerns a different LLVM machine backend so it would probably have to be addressed separately.

**Comment by tatetian** [comment]
@workingjubilee My apologies; I may have given the wrong impression. I shared the situation we encountered in Asterinas to provide additional context for the discussion, but **my personal judgment is that this is not an LLVM or rustc problem**. I largely agree with your assessment:

> The basic contract of `write_volatile` and `read_volatile` is to "work with MMIO." The problem is that **it does work with MMIO**. Does it work with MMIO *and* a hypervisor *and* protected mode? That is a separate "composing of features" problem.

Asterinas is an OS kernel. As kernel developers, we are comfortable writing assembly when necessary, and performing MMIO via instruction emulation within Confidential VMs (CVMs) falls squarely into that category. LLVM has no way of knowing which instructions our kernel can decode or emulate. In general, **LLVM cannot—and should not—assume which instructions are safe to emit for a protected/confidential setting**, as that depends entirely on the specific kernel/hypervisor implementation.

Furthermore, if one argues that LLVM or Rust should natively support MMIO in CVM settings, where do we draw the line? For instance, would the LLVM intrinsic `volatile_copy_memory` also need to be modified to support CVM-specific constraints?

@joshtriplett I am afraid that **this cannot be resolved via instruction selection. At build time, the compiler cannot determine whether the binary will run in a standard environment or a Confidential VM**. Asterinas (much like Rust for Linux) is a general-purpose kernel; it *supports* CVMs but is not exclusively deployed in them. Consequently, the kernel must detect its environment at boot time and _dynamically_ select the appropriate implementation for MMIO reads, writes, and copies to ensure both correctness and efficiency.

Ultimately, **addressing this at the library or kernel level is both simple and efficient**—provided one is comfortable with inline assembly:

```rust
/// Reads from a pointer using a non-tearing memory load.
///
/// This function is semantically equivalent to `core::ptr::read_volatile`
/// but is manually implemented with a single memory load instruction.
/// If this instruction faults inside a confidential VM, it can be 
/// decoded and emulated by the kernel.
pub(crate) unsafe fn read_once<T: PodOnce>(ptr: *const T) -> T {
    let mut val: u64 = 0;

    // SAFETY: The caller must ensure `ptr` is valid.
    unsafe {
        // EFFICIENCY: The match is optimized out in release builds.
        match core::mem::size_of::<T>() {
            1 => asm!("mov {0:l}, [{1}]", out(reg) val, in(reg) ptr, options(nostack, readonly, preserves_flags)),
            2 => asm!("mov {0:x}, [{1}]", out(reg) val, in(reg) ptr, options(nostack, readonly, preserves_flags)),
            4 => asm!("mov {0:e}, [{1}]", out(reg) val, in(reg) ptr, options(nostack, readonly, preserves_flags)),
            8 => asm!("mov {0}, [{1}]",   out(reg) val, in(reg) ptr, options(nostack, readonly, preserves_flags)),
            _ => core::hint::unreachable_unchecked(),
        }
        // EFFICIENCY: This compiles to a no-op in release builds.
        core::ptr::read(core::ptr::addr_of!(val) as *const T)
    }
}

/// Plain-old data that supports `read_once`/`write_once`.
pub unsafe trait PodOnce {}

// Implement PodOnce for u8, u16, u32, u64, usize and their signed counterparts.
// ...
```

While I am speaking from the perspective of the Asterinas project, I believe **these points apply to many other Rust-based kernels or embedded systems**.

**Comment by mkroening** [comment]
> Unless the compiler knows for sure that your binary will not run under a Confidential VM, it should not use any of the few instructions that are impossible for a Confidential VM to execute, because it might run under a Confidential VM. Emitting an instruction that doesn't _and can't_ do the right thing under some circumstances is simply miscompilation.

To give this some more context:

Handling other memory access operations apart from `mov` such as `cmp` is generally possible to do in AMD SEV kernels. That makes the instruction decoder somewhat larger, of course, but that is what @zyuiop has done for Hermit (not public yet).

As far as we know, both OVMF ([OvmfPkg/Library/CcExitLib/CcExitVcHandler.c#L147-L167](https://github.com/tianocore/edk2/blob/edk2-stable202511/OvmfPkg/Library/CcExitLib/CcExitVcHandler.c#L147-L167)) and Linux ([arch/x86/coco/sev/vc-handle.c#L578](https://github.com/torvalds/linux/blob/v6.19/arch/x86/coco/sev/vc-handle.c#L578)) only handle `mov` instructions, though, since they use assembly for accessing MMIO memory instead of C's `volatile` keyword.

I think it is not necessarily wrong to optimize `volatile_read(ptr) == 0` from `mov` + `cmp` to `cmp` ([Compiler Explorer of the C equivalent](https://godbolt.org/z/hKxfr5WP1)), since this _can_ be handled correctly. Interestingly, this optimization is only done by Clang and not by GCC. It might make sense to check whether GCC does not do this on purpose and to check LLVM's stance on this. Personally, I have no strong opinion on whether this optimization should be done or not.

**Comment by workingjubilee** [comment]
This has been addressed upstream for aarch64! https://github.com/llvm/llvm-project/pull/196305

**Comment by workingjubilee** [comment]
For any x86(-64) stuff, please open a new issue.
